File: src/main/java/application/RayTabController.java
package application;

import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.ColorPicker;
import javafx.scene.control.TextField;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import optics.PreciseLine;
import optics.light.Ray;

import static application.Storage.*;

public class RayTabController {

  @FXML
  private Button newRay;
  @FXML
  private TextField rayRotation;
  @FXML
  private ColorPicker rayColor;

  private Ray focusedRay;
  private String expectedText;
  private Color expectedColor;

  public void initialize(Pane parent) {
    newRay.setOnMouseClicked(e -> {
      // Prevent changes when animating
      if (Storage.isAnimating) return;
      PreciseLine l = new PreciseLine(parent.getWidth() / 2, parent.getHeight() / 2, parent
              .getWidth() / 2 + 2500, parent.getHeight() / 2);
      l.setPreciseAngle(0);
      Ray r = new Ray(l, parent);
      this.createRay(r);
    });
    rayRotation.textProperty().addListener((o, ol, val) -> {
      // Prevent changes when animating
      if (Storage.isAnimating) return;
      if (val.equals(expectedText)) return;
      if (val.length() == 0) {
        if (this.focusedRay == null) return;
        this.focusedRay.setAngle(0);
        rerenderRay(this.focusedRay);
        return;
      }
      double value;
      try {
        value = Double.parseDouble(val);
      } catch (NumberFormatException e) {
        return;
      }
      if (this.focusedRay == null) return;
      this.focusedRay.setAngle(Double.parseDouble(fixAngle(value)));
      rerenderRay(this.focusedRay);
    });
    rayColor.valueProperty().addListener((o, ol, color) -> {
      // Prevent changes when animating
      if (Storage.isAnimating) return;
      if (this.focusedRay == null) return;
      if (color.equals(expectedColor)) return;
      this.focusedRay.setColor(color);
      this.focusedRay.renderRays(opticalRectangles);
    });
  }

  public void createRay(Ray r, boolean syncToServer) {
    changeFocus(r);
    rays.add(r);
    r.setOnDestroy(e -> {
      rays.remove(r);
      return null;
    });
    r.addOnStateChange(e -> {
      changeFocus(r);
      rerenderRay(r);
    });
    r.setOnFocusStateChanged(state -> {
      if (state) changeFocus(r);
      return null;
    });
    rerenderRay(r);
    this.expectedText = fixAngle(r.getAngle());
    this.focusedRay = r;
    this.expectedColor = Color.BLACK;
    rayRotation.setText(expectedText);
    rayColor.valueProperty().setValue(Color.BLACK);
    r.requestFocus();
  }

  public void createRay(Ray r) {
    createRay(r, false);
  }

  private void changeFocus(Ray r) {
    if (this.focusedRay != null) {
      this.focusedRay.getCircle().setStroke(Color.BLACK);
      this.focusedRay.getCircle().setStrokeWidth(1);
    }
    this.focusedRay = r;
    this.focusedRay.getCircle().setStroke(Color.BLUE);
    this.focusedRay.getCircle().setStrokeWidth(2);
    this.expectedText = fixAngle(r.getAngle());
    this.expectedColor = this.focusedRay.getColor();
    rayColor.setValue(this.focusedRay.getColor());
    rayRotation.setText(expectedText);
  }

  Ray getFocusedRay() {
    return focusedRay;
  }

  private String fixAngle(double angle) {
    angle %= 360;
    if (angle < 0) angle += 360;
    return String.valueOf(angle);
  }
}

File: src/main/java/application/RayTabController.kt
package application

import application.Storage.rays
import application.Storage.reRenderAll
import application.Storage.rerenderRay
import javafx.beans.value.ObservableValue
import javafx.event.Event
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.geometry.Point2D
import javafx.scene.control.Button
import javafx.scene.control.ColorPicker
import javafx.scene.control.TextField
import javafx.scene.input.MouseEvent
import javafx.scene.layout.Pane
import javafx.scene.paint.Color
import optics.RealLine
import optics.light.Ray
import utils.toDegrees

class RayTabController {
    @FXML
    private var newRay: Button? = null

    @FXML
    private var rayRotation: TextField? = null

    @FXML
    private var rayColor: ColorPicker? = null
    var focusedRay: Ray? = null
        private set
    private var expectedText: String? = null
    private var expectedColor: Color? = null
    fun initialize(parent: Pane) {
        newRay!!.onMouseClicked = EventHandler { e: MouseEvent? ->
            // Prevent changes when animating
            if (Storage.isAnimating) return@EventHandler
            val l = RealLine(Point2D(parent.width / 2, parent.height / 2),
                Point2D(parent.width / 2 + 25000, parent.height / 2))
            val r = Ray(l, Color.BLACK, parent)
            createRay(r)
        }
        rayRotation!!.textProperty()
            .addListener { o: ObservableValue<out String>?, ol: String?, strVal: String ->
                // Prevent changes when animating
                if (Storage.isAnimating) return@addListener
                if (strVal == expectedText) return@addListener
                val ray = focusedRay ?: return@addListener
                if (strVal.isEmpty()) {
                    val newRealLine = ray.realLine.copyWithAngle(0.0)
                    ray.update(newRealLine)
                    return@addListener
                }
                val value: Double = try {
                    strVal.toDouble()
                } catch (e: NumberFormatException) {
                    return@addListener
                }
                val newRealLine = ray.realLine.copyWithAngle(value)
                ray.update(newRealLine)
            }
        rayColor!!.valueProperty()
            .addListener { _: ObservableValue<out Color>?, _: Color?, color: Color ->
                // Prevent changes when animating
                if (Storage.isAnimating) return@addListener
                val ray = focusedRay ?: return@addListener
                if (color == expectedColor) return@addListener
                ray.update(newColor = color)
            }
    }

    fun createRay(r: Ray) {
        changeFocus(r)
        rays.add(r)
        r.setOnDestroy {
            rays.remove(r)
        }
        r.addOnStateChange {
            changeFocus(it)
            rerenderRay(it)
        }
        r.setOnFocusStateChanged { state: Boolean ->
            if (state) changeFocus(r)
        }
        expectedText = fixAngle(r.realLine.angle.toDegrees())
        focusedRay = r
        expectedColor = Color.BLACK
        rayRotation!!.text = expectedText
        rayColor!!.valueProperty().value = Color.BLACK
        rerenderRay(r)
        r.requestFocus()
    }

    private fun changeFocus(r: Ray) {
        if (focusedRay != null) {
            focusedRay!!.circle.stroke = Color.BLACK
            focusedRay!!.circle.strokeWidth = 1.0
        }
        focusedRay = r
        r.circle.stroke = Color.BLUE
        r.circle.strokeWidth = 2.0
        expectedText = fixAngle(r.realLine.angle.toDegrees())
        expectedColor = r.color
        rayColor!!.value = r.color
        rayRotation!!.text = expectedText
    }

    private fun fixAngle(angle: Double): String {
        var angle = angle
        angle %= 360.0
        if (angle < 0) angle += 360.0
        return angle.toString()
    }
}
